src/evo/deterministicmns.h

```diff
@@ -0,0 +1,441 @@
+        DeleteUniqueProperty(dmn, dmn->state->keyIDOwner);
+        DeleteUniqueProperty(dmn, dmn->state->keyIDOperator);
+        mnMap = mnMap.erase(proTxHash);
+    }
+
+    template<typename T>
+    bool HasUniqueProperty(const T& v) const
+    {
+        return mnUniquePropertyMap.count(::SerializeHash(v)) != 0;
+    }
+    template<typename T>
+    CDeterministicMNCPtr GetUniquePropertyMN(const T& v) const
+    {
+        auto p = mnUniquePropertyMap.find(::SerializeHash(v));
+        if (!p) {
+            return nullptr;
+        }
+        return GetMN(p->first);
+    }
+
+private:
+    bool IsMNValid(const CDeterministicMNCPtr& dmn) const;
+    bool IsMNPoSeBanned(const CDeterministicMNCPtr& dmn) const;
+
+    template<typename T>
+    void AddUniqueProperty(const CDeterministicMNCPtr& dmn, const T& v)
+    {
+        auto hash = ::SerializeHash(v);
+        auto oldEntry = mnUniquePropertyMap.find(hash);
+        assert(!oldEntry || oldEntry->first == dmn->proTxHash);
+        std::pair<uint256, uint32_t> newEntry(dmn->proTxHash, 1);
+        if (oldEntry) {
+            newEntry.second = oldEntry->second + 1;
+        }
+        mnUniquePropertyMap = mnUniquePropertyMap.set(hash, newEntry);
+    }
+    template<typename T>
+    void DeleteUniqueProperty(const CDeterministicMNCPtr& dmn, const T& oldValue)
+    {
+        auto oldHash = ::SerializeHash(oldValue);
+        auto p = mnUniquePropertyMap.find(oldHash);
+        assert(p && p->first == dmn->proTxHash);
+        if (p->second == 1) {
+            mnUniquePropertyMap = mnUniquePropertyMap.erase(oldHash);
+        } else {
+            mnUniquePropertyMap = mnUniquePropertyMap.set(oldHash, std::make_pair(dmn->proTxHash, p->second - 1));
+        }
+    }
+    template<typename T>
+    void UpdateUniqueProperty(const CDeterministicMNCPtr& dmn, const T& oldValue, const T& newValue)
+    {
+        if (oldValue == newValue) {
+            return;
+        }
+        DeleteUniqueProperty(dmn, oldValue);
+        AddUniqueProperty(dmn, newValue);
+    }
+};
+
+class CDeterministicMNListDiff
+{
+public:
+    uint256 prevBlockHash;
+    uint256 blockHash;
+    int height{-1};
+    std::map<uint256, CDeterministicMNCPtr> addedMNs;
+    std::map<uint256, CDeterministicMNStateCPtr> updatedMNs;
+    std::set<uint256> removedMns;
+
+public:
+    ADD_SERIALIZE_METHODS;
+
+    template <typename Stream, typename Operation>
+    inline void SerializationOp(Stream& s, Operation ser_action)
+    {
+        READWRITE(prevBlockHash);
+        READWRITE(blockHash);
+        READWRITE(height);
+        READWRITE(addedMNs);
+        READWRITE(updatedMNs);
+        READWRITE(removedMns);
+    }
+
+public:
+    bool HasChanges() const
+    {
+        return !addedMNs.empty() || !updatedMNs.empty() || !removedMns.empty();
+    }
+};
+
+class CDeterministicMNManager
+{
+    static const int SNAPSHOT_LIST_PERIOD = 576; // once per day
+    static const int LISTS_CACHE_SIZE = 576;
```

Comment:
This was my initial plan, but then I realized that someone might be able to come up with a way to rotate the cache multiple times in a single block, causing `GetListForBlock` to not use any cached entries and thus slow down block processing. Not sure if this would be really possible as of now, but if we ever introduce some special transaction that accesses old lists this might become an attack vector. So I preferred to use a caching mechanism that did a cleanup after block processing.

Just checked the implementation of CacheMap from cachemap.h. If we add a boolean template parameter to this map that disables automatic pruning and instead allows manual pruning, we could use that map implementation for this kind of "cleanup" behavior. What do you think?